cmd/node/main.go

```diff
@@ -85,12 +73,132 @@ func main() {
+		controller.NewGetNodeRoute(db.Manager),
+		controller.NewAddUserRoute(nodeCtrl),
+		controller.NewInstallAppRoute(nodeCtrl),
+		controller.NewStartProcessHandler(nodeCtrl),
+		controller.NewMigrationRoute(nodeCtrl),
+	}
+
+	router := api.NewRouter(routes, log, nodeCtrl, nodeConfig)
+	apiServer := &http.Server{
+		Addr:        fmt.Sprintf(":%s", constants.DefaultPortHabitatAPI),
+		TLSConfig:   tlsConfig,
+		Handler:     router,
+		BaseContext: func(net.Listener) context.Context { return context.Background() },
+	}
+	url, err := url.Parse(fmt.Sprintf("http://localhost:%s", constants.DefaultPortHabitatAPI))
+	if err != nil {
+		log.Fatal().Err(fmt.Errorf("error parsing Habitat API URL: %v", err))
+	}
+	err = proxy.Rules.Add("Habitat API", &reverse_proxy.RedirectRule{
+		ForwardLocation: url,
+		Matcher:         "/habitat/api",
+	})
+	if err != nil {
+		log.Fatal().Err(fmt.Errorf("error adding Habitat API proxy rule: %v", err))
+	}
+	eg.Go(serveFn(apiServer, "api-server", WithTLSConfig(nodeConfig.NodeCertPath(), nodeConfig.NodeKeyPath())))
+
+	// Wait for either os.Interrupt which triggers ctx.Done()
+	// Or one of the servers to error, which triggers egCtx.Done()
+	select {
+	case <-egCtx.Done():
+		log.Err(egCtx.Err()).Msg("sub-service errored: shutting down Habitat")
+	case <-ctx.Done():
+		log.Info().Msg("Interrupt signal received; gracefully closing Habitat")
+		stop()
+	}
+
+	// Shutdown the API server
+	err = apiServer.Shutdown(context.Background())
+	if err != nil {
+		log.Err(err).Msg("error on api-server shutdown")
+	}
+	log.Info().Msg("Gracefully shutdown Habitat API server")
+
+	// Shutdown the proxy server
+	err = proxyServer.Shutdown(context.Background())
+	if err != nil {
+		log.Err(err).Msg("error on proxy-server shutdown")
+	}
+	log.Info().Msg("Gracefully shutdown Habitat proxy server")
+
+	// Wait for the go-routines to finish
+	err = eg.Wait()
+	if err != nil {
+		log.Err(err).Msg("received error on eg.Wait()")
+	}
+	log.Info().Msg("Finished!")
+}
+
+// config provided to http.Server.ListenAndServeTLS()
+type tlsConfig struct {
+	certFile string
+	keyFile  string
+}
+
+// serverOptions provide optional config for an http.Server passed to serveFn()
+type serverOption struct {
+	tlsConfig *tlsConfig
+}
+
+// conventional way to supply arbitrary and optional arguments as options to a function.
+type option func(*serverOption)
+
+// WithTLSConfig provides serverOption with tlsConfig
+func WithTLSConfig(certFile string, keyFile string) option {
+	return func(so *serverOption) {
+		so.tlsConfig = &tlsConfig{
+			certFile: certFile,
+			keyFile:  keyFile,
+		}
+	}
+}
+
+// serveFn takes in an http.Server and additional config and returns a callback that can be run in a separate go-routine.
+func serveFn(srv *http.Server, name string, opts ...option) func() error {
+	options := &serverOption{}
+	for _, o := range opts {
+		o(options)
+	}
+	return func() error {
+		var err error
+		if srv.TLSConfig != nil && options.tlsConfig != nil {
```

Comment:
Let's modify this to refer to `USE_TSNET` once this is merged: https://github.com/eagraf/habitat-new/pull/16/commits/9cae3e6efc4cdda0e6b96cc52b099773260620d7